todos#index の実装

chionyan opened the pull request:
## 概要
### TodoSerializer の実装
- [Simple TODO API の仕様](https://app.swaggerhub.com/apis/kielze/TODO-API/1.0.0)の JSON 形式でデータを返すように設定
- `created_at` については、 https://github.com/chionyan/simple-todo-api/pull/6#pullrequestreview-146309551 に記載
<img width="1291" alt="2018-08-15 10 45 22" src="https://user-images.githubusercontent.com/26681827/44127469-5d3817aa-a078-11e8-85ff-a6d8275abceb.png">

### todos#index の実装
- ルーティングの設定
- 取得したデータを JSON 形式の配列で render する
- Request Spec の作成
  - HTTP Status 200 がレスポンスされるか
  - `todos` データ配列が JSON 形式でレスポンスされるか

## 動作確認
- [ ] Rails
  - `$ bin/setup` を実行
  - `$ bundle exec rails s` で Rails サーバーを起動
  - [ ] 他ターミナルで下記コマンドを実行し、3件の JSON 形式のデータが返ってくることを確認

  ```sh
  $ curl -X GET "http://localhost:3000/todos" -H "accept: application/json"
  [
    {"id":"440b1af7-52a5-47c2-81a9-865731a32b69","title":"title_0","text":"text_0","created_at":"2018-08-13T09:27:48Z"},
    {"id":"25d1daec-1afc-4e89-bb8e-7d37c57d30cc","title":"title_1","text":"text_1","created_at":"2018-08-13T09:27:48Z"},
    {"id":"d081f498-64dd-4629-a66b-d1d7f0523a00","title":"title_2","text":"text_2","created_at":"2018-08-13T09:27:48Z"}
  ]
  ```

- [ ] RSpec
  - `$ bundle exec rspec` を実行
  - [ ] `0 failures` が出力されることを確認

- [ ] Rubocop
  - `$ bundle exec rubocop` を実行  
  - [ ] `no offenses detected` が出力されることを確認

chionyan committed:
$ rails g controller Todos

chionyan committed:
$ rails g rspec:integration Todo

chionyan committed:
Describe Todo#index, Method, Routing

chionyan committed:
Merge branch 'master' into feature-todo-index

chionyan committed:
Add created_at sort

chionyan committed:
$ rails g serializer todo

- TodoSerializer の created_at を iso8601 に設定
- Todo Factory の作成
- Request Spec に JSON 形式でデータが返ってくるテストを実装

chionyan commented on app/serializers/todo_serializer.rb:
```diff
@@ -0,0 +1,7 @@
+
+  def created_at
+    object.created_at&.in_time_zone&.iso8601
+  end
```
[Simple TODO API の仕様](https://app.swaggerhub.com/apis/kielze/TODO-API/1.0.0)に記載されているように、`created_at` を ISO8601 形式の文字列にフォーマットするようにしています。

koshigoe replied:
🙏 参考までに、 `in_time_zone` を利用している意図を教えてください。

```
$ bin/setup
$ rails c
> Todo.first.created_at.iso8601
  Todo Load (0.5ms)  SELECT  "todos".* FROM "todos" ORDER BY "todos"."id" ASC LIMIT $1  [["LIMIT", 1]]
=> "2018-08-15T02:45:35Z"
> Todo.first.created_at.in_time_zone.iso8601
  Todo Load (0.4ms)  SELECT  "todos".* FROM "todos" ORDER BY "todos"."id" ASC LIMIT $1  [["LIMIT", 1]]
=> "2018-08-15T02:45:35Z"
> Todo.first.created_at.zone
  Todo Load (0.4ms)  SELECT  "todos".* FROM "todos" ORDER BY "todos"."id" ASC LIMIT $1  [["LIMIT", 1]]
=> "UTC"
> Time.zone
=> #<ActiveSupport::TimeZone:0x00007fab49ab3aa0 @name="UTC", @tzinfo=#<TZInfo::DataTimezone: Etc/UTC>, @utc_offset=nil>
> Time.current.in_time_zone('Asia/Tokyo').iso8601
=> "2018-08-15T11:48:32+09:00"
```

chionyan replied:
(口頭説明)
`ActiveSupport::TimeWithZone#in_time_zone` は引数に与えたタイムゾーンに変更する際に使用する。
参考：[ActiveSupport::TimeWithZone\#in\_time\_zone](https://www.rubydoc.info/docs/rails/3.2.8/ActiveSupport%2FTimeWithZone:in_time_zone)
そのため、ここでは使用する必要がないと考えられる。また、[Time\#iso8601](https://docs.ruby-lang.org/ja/latest/method/Time/i/iso8601.html)は Time Class に使用できる。

---
[ISO 8601 のWiki](https://ja.wikipedia.org/wiki/ISO_8601#%E6%97%A5%E4%BB%98%E3%81%A8%E6%99%82%E5%88%BB%E3%81%AE%E7%B5%84%E5%90%88%E3%81%9B) に

 > 時刻の後ろに Z を添えることで協定世界時（UTC）での時刻をそのまま示すことができる。

とあるので、タイムゾーンは UTC を使用しました！

chionyan commented on app/controllers/todos_controller.rb:
```diff
@@ -0,0 +1,6 @@
+class TodosController < ApplicationController
+  def index
+    todos = Todo.all.order(:created_at)
```
意図しない順序でデータが返ってくるのを避けるため、 `created_at` でソートしてデータを返しています。
参考：[PostgreSQLでORDER BY区を指定しない場合の並び](http://docs.komagata.org/5169)

koshigoe replied:
❓仕様について確認です。

- ページネーションなしの全件出力で正しいですよね？
- 作成日時昇順(古い順)で並べるという事で正しいですよね？

koshigoe replied:
口頭でその通りである旨、確認済み。

chionyan committed:
Change Time.in_time_zone.iso8601 -> Time&.in_time_zone&.iso8601

chionyan commented on spec/requests/todos_spec.rb:
```diff
@@ -10,7 +10,7 @@
         'title' => todo.title,
         'text' => todo.text,
-        'created_at' => todo.created_at.in_time_zone.iso8601,
+        'created_at' => todo.created_at&.in_time_zone&.iso8601,
```
`todo.created_at` が `nil` の可能性もあるため、`&.` 演算子に変更しました。

chionyan commented:
@koshigoe @pokotyamu
Todo#index の実装を一通り作成しました！レビューよろしくお願いします🙏✨

koshigoe commented:
👀

koshigoe commented:
『「何」を実現するため』に概要に書かれている事を実施したのかを書いてもらえると、PR の目的が達成されているか判断しやすくて助かります。🙏

🤔 以下の認識であってますか？
敢えてこの PR では行っていない(スコープ外とした)事があれば、改めて教えてください。 🙏 

目的(ゴール)
----

ToDo 一覧を取得するための API を実装する。

API 仕様
----

ref. https://app.swaggerhub.com/apis/kielze/TODO-API/1.0.0#/developers/searchTodo

```
GET /todos[?searchString=<String>]

[
  {
    "id": "d290f1ee-6c54-4b01-90e6-d701748f0851",
    "title": "Cook dinner",
    "text": "Today we will cook pork curry.\nSo please buy these food ingredients, potato, carrot, onion, and pork meat.\nAhh, may be we are out of garam masala, check it before go out.\n",
    "created_at": "2018-06-04T08:10:06Z"
  }
]
```

### リクエスト

- リクエストパラメータ `searchString` で検索可能？

### レスポンス

- レスポンスボディのフォーマットは JSON 固定(選択不可)
- 正常応答(200 OK)
    - ToDo オブジェクトの配列を JSON フォーマットで返す
    - ToDo オブジェクトのデータ構造は後述
    - ToDo オブジェクトの並び順は、日付昇順？(古い順)
    - ページネーション機能はない(全件出力)
- エラー応答(4xx, 5xx)
    - エラーデータの配列を JSON フォーマットで返す(`{"errors":<Array>}`)
    - エラーデータのデータ構造は `{"title":<String>,"status":<Integer>}`
    - エラー条件未定義？

| 項目名       | 型       | 備考                                                |
| ------------ | -------- | --------------------------------------------------- |
| `id`         | `string` | `Todo#id` (UUID) を文字列にしたもの                 |
| `title`      | `string` | `Todo#title` そのまま                               |
| `text`       | `string` | `Todo#text` そのまま                                |
| `created_at` | `string` | `Todo#created_at` を UTC の ISO 8601 形式にしたもの |

koshigoe commented on config/routes.rb:
```diff
@@ -1,3 +1,4 @@
 Rails.application.routes.draw do
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+  resources :todos, only: [:index]
 end
```
👀  API レスポンスを確認。

```
$ bin/setup
$ rails s
```

```
$ curl -s http://localhost:3000/todos | jq .
[
  {
    "id": "c656b14e-ed86-4bab-80ac-74b6af9c6b87",
    "title": "title_0",
    "text": "text_0",
    "created_at": "2018-08-15T02:51:27Z"
  },
  {
    "id": "6d24caa9-eafd-4bfc-a479-30f135b60c17",
    "title": "title_1",
    "text": "text_1",
    "created_at": "2018-08-15T02:51:27Z"
  },
  {
    "id": "7840d40c-2e80-4db2-b682-e04721b9b914",
    "title": "title_2",
    "text": "text_2",
    "created_at": "2018-08-15T02:51:27Z"
  }
]
```

SQL 的に `ORDER BY created_at ASC` である事も確認。

```
Started GET "/todos" for 127.0.0.1 at 2018-08-15 11:52:18 +0900
Processing by TodosController#index as */*
  Todo Load (6.3ms)  SELECT "todos".* FROM "todos" ORDER BY "todos"."created_at" ASC
  ↳ app/controllers/todos_controller.rb:4
[active_model_serializers] Rendered ActiveModel::Serializer::CollectionSerializer with ActiveModelSerializers::Adapter::Attributes (1.56ms)
Completed 200 OK in 11ms (Views: 4.1ms | ActiveRecord: 6.3ms)
```

koshigoe commented on spec/requests/todos_spec.rb:
```diff
@@ -0,0 +1,40 @@
+
+RSpec.describe 'Todos', type: :request do
+  describe 'GET /todos' do
+    subject { get todos_path }
```
(ただの感想です。)

💭 `todos_path` を使っても良いと思う反面、テストコードを見ただけですぐにリクエストパスが分かった方が良い気もする。自分はパスをべた書きする流儀。

chionyan replied:
d060619 で直のパスを get するように修正しました！

メモ：
先ほどお話しした内容から、今後の `post`  `update` では `params` を渡すことを想定されるのと、 describe のテキストと同一の内容を subject 内で行なっているため、 Request Spec における主題のテストは subject 内に記述するようにします。

koshigoe commented on spec/requests/todos_spec.rb:
```diff
@@ -0,0 +1,40 @@
+        subject
+        expect(JSON.parse(response.body)[1]).to eq expect_todo_result
+      end
+    end
```
💭`context` を使って Todo 要素単位でテストケースを分割するのは、たぶんやりすぎ。
💭ひとつの `example` の中で期待する配列との一致を比較した方が分かりやすいというのが、自分の流儀。
💭要素単位で比較するのでも良いけど、そうするなら要素数の確認は必須。

chionyan replied:
97b29d4 で返ってきた JSON の要素数を確認するテストを追加しました！

koshigoe commented on spec/requests/todos_spec.rb:
```diff
@@ -0,0 +1,40 @@
+        'text' => todo.text,
+        'created_at' => todo.created_at&.in_time_zone&.iso8601,
+      }
+    end
```
💭テストケースから離れた(切り出された)場所で期待値を動的に組み立てる手法は、個人的にテストを分かりにくくする要因だと考えている。
💭自分だったらという話になるけど、期待値は極力 `example` の中で愚直に書く。(可能な限りリテラルで)

```ruby
expected = [
  {...},
  ...
]
res = JSON.parse(response.body)
expect(res).to eq expected
```

koshigoe replied:
❓ちなみに、テストコードの期待値でも `in_time_zone` とか `&.` って必要なんですか？

chionyan replied:
テストでは `create` に対して期待していない値が入っている場合はテストが落ちるべきなので、 `&.` ではなく、 `.` に変更しました！

chionyan commented:
@koshigoe 
コメントいただいた内容を元に、今回のPRでの実装の目的、範囲などを下記に記述してますので、目を通していただけるとと思います！
以後の実装は、PRでの目的、PRでの範囲外の内容などを0コメに記述するように意識します🙇‍♀️

目的(ゴール)
----

ToDo 一覧を取得するための API を実装する。

API 仕様
----

ref. https://app.swaggerhub.com/apis/kielze/TODO-API/1.0.0#/developers/searchTodo

```
GET /todos[?searchString=<String>]

[
  {
    "id": "d290f1ee-6c54-4b01-90e6-d701748f0851",
    "title": "Cook dinner",
    "text": "Today we will cook pork curry.\nSo please buy these food ingredients, potato, carrot, onion, and pork meat.\nAhh, may be we are out of garam masala, check it before go out.\n",
    "created_at": "2018-06-04T08:10:06Z"
  }
]
```

### レスポンス
- レスポンスボディのフォーマットは JSON 固定(選択不可)
- 正常応答(200 OK)
    - ToDo オブジェクトの配列を JSON フォーマットで返す
    - ToDo オブジェクトの並び順は日付昇順
    - ページネーション機能はない(全件出力)

-  ToDo オブジェクトのデータ構造

| 項目名       | 型       | 備考                                                |
| ------------ | -------- | --------------------------------------------------- |
| `id`         | `string` | `Todo#id` (UUID) を文字列にしたもの                 |
| `title`      | `string` | `Todo#title` そのまま                               |
| `text`       | `string` | `Todo#text` そのまま                                |
| `created_at` | `string` | `Todo#created_at` を UTC の ISO 8601 形式にしたもの |

### このPRでのスコープ外
- エラー応答(400)のエラーハンドリング
    - エラーデータの配列を JSON フォーマットで返す
    - エラーデータのデータ構造は `{"errors": [{"title":<String>,"status":<Integer>}]}`
    - エラー条件は[WIP]

### 実装全体におけるスコープ外
- リクエストパラメータ `searchString` で検索可能にする（時間的余裕があれば）
- サーバエラー応答(5xx)のエラーハンドリング


また、このPRでのスコープ外ではありますが、エラー応答(400)の条件がまだ見えていないので、そちらについてはエラー応答時のPRの際に記述します！

---

@koshigoe さんの認識からの変更点
- エラーハンドリング、リクエストパラメータの実装はこのPR（または実装）からスコープ外にしています！

koshigoe reviewed (commented):
@chionyan コメントしました。

勢いに任せてずらーっとこめんとしたので、よくわからんかったら適当なタイミング(いつでもok)で声かけてもらえれば説明します。

koshigoe commented:
> コメントいただいた内容を元に、今回のPRでの実装の目的、範囲などを下記に記述してますので、目を通していただけるとと思います！

拝承。
わたしが想像で書いた内容に対する訂正箇所も添えて貰えると確認が捗ります。 🙏🙏🙏🙏🙏

chionyan committed:
Fix created_at.in_time_zone.iso8601 -> created_at&.iso8601

chionyan committed:
Using travel_to

chionyan committed:
Change get todos_path -> get /todos

chionyan committed:
Use literal parameter

- Todo Factory で sequence を使わず静的な文字列で定義する
- テストで想定される結果も create 時に与えた文字列や travel_to で固定した時間を使用する

chionyan committed:
Add JSON Count Test

chionyan commented:
コメントや口頭で説明いただいたように、 spec ファイル内で何をテストしたいか明確にわかるように、 `it` 内で想定される結果をリテラルを用いて記述しています！
※ created_at については aed8758 で `travel_to` を使用するようにし、作成される日時を固定するようにしています。
※ id については uuid を使用しており動的に作成されるため、そこだけ変数から取得するようにしています。

chionyan commented:
@koshigoe @pokotyamu 
現時点でのご指摘いただいた内容を修正いたしました！再度レビューよろしくお願いします🙇‍♀️💦

koshigoe commented:
👀

koshigoe commented on spec/factories/todo.rb:
```diff
@@ -0,0 +1,6 @@
+    title 'Sample title'
+    text 'Sample text'
+  end
+end
```
💭 factory 的には sequence を使っていて問題なかったと思うけど、無くても困らない。

chionyan replied:
確かにそれは思いました🙁
いずれ実装する `Todo#show` のテストのことを考えると、固定値の方が書きやすいかな、と思って固定値にしておきました…！

koshigoe commented on spec/requests/todos_spec.rb:
```diff
@@ -0,0 +1,42 @@
+    it 'return 2 JSON' do
+      subject
+      expect(JSON.parse(response.body).count).to eq 2
+    end
```
(重箱の隅の感想なので気にしなくて良いです)

💭 自分だったら不必要に example を分割したくないので、全部一つにまとめてしまう。
💭 分けるにしても、ヘッダとボディで分けるくらいで、レスポンスボディの検証についてはひとまとめにする。
💭 `return valid JSON` で、一度で済む `JSON.parse` を繰り返し実行しているのは無駄だなと感じる。

chionyan replied:
英語力の問題ですが、 valid が「妥当な」という意味なのを初めて知りました😱
したがって、「妥当な JSON（ボディ）」という意味合いだと、 返ってくるデータの数を確認するのもどういるの `it` に含めるべきだと思ったので、 81354b4 で修正しております😀

koshigoe replied:
👁 みた。👍

koshigoe reviewed (approved):
言いたいことは言えたので、わたしは満足です。

chionyan committed:
Change Test Structure

pokotyamu commented on spec/requests/todos_spec.rb:
```diff
@@ -0,0 +1,40 @@
+
+      expect(result_todos.count).to eq 2
+      expect(result_todos[0]).to eq expect_todo_first
+      expect(result_todos[1]).to eq expect_todo_second
```
(修正の必要はなし)
１つの `it` の中で、`expect` が複数あるときは、 `aggregate_failures` を使うと、途中で失敗しても最後までテストしてくれるので、オススメですよ 😄

chionyan replied:
確かにテストをまとめた時にその点を懸念していたので、 47759ae で  `aggregate_failures` でくくっておきました！今後も使いそうなので😆

pokotyamu commented on spec/rails_helper.rb:
```diff
@@ -73,4 +73,7 @@
   config.include FactoryBot::Syntax::Methods
+
+  config.include ActiveSupport::Testing::TimeHelpers
+  config.after { travel_back }
```
`travel_to` の戻り忘れはハマりどころだと思うので、自分も結構調べたりしています 🙏

chionyan replied:
(口頭説明)
io では `around` の中で `travel_to` を使っていることが多い
すると、rails_helper.rb で `config.after { travel_back }` を記述する必要がない

pokotyamu reviewed (approved):
よさみ 👍

chionyan committed:
Add aggregate_failures block
